vms/evm/subscriber.go

```diff
@@ -84,7 +117,73 @@ func (s *subscriber) forwardLogs() {
 			continue
 		}
 		s.log <- *messageInfo
+
+		// Update the database with the latest seen block height
+		err = s.db.Put(s.chainID, []byte(database.LatestSeenBlockKey), []byte(strconv.FormatUint(msgLog.BlockNumber, 10)))
+		if err != nil {
+			s.logger.Error(fmt.Sprintf("failed to put %s into database", database.LatestSeenBlockKey), zap.Error(err))
+		}
+	}
+}
+
+func (s *subscriber) ProcessFromHeight(height *big.Int) error {
+	ethClient, err := ethclient.Dial(s.nodeRPCURL)
+	if err != nil {
+		return err
+	}
+
+	// Grab the latest block before filtering logs so we don't miss any before updating the db
+	latestBlock, err := ethClient.BlockNumber(context.Background())
+	if err != nil {
+		s.logger.Error(
+			"Failed to get latest block",
+			zap.Error(err),
+		)
+		return err
+	}
+
+	// Filter logs from the latest processed block to the latest block
+	// Since initializationFilterQuery does not modify existing fields of warpFilterQuery,
+	// we can safely reuse warpFilterQuery with only a shallow copy
+	initializationFilterQuery := interfaces.FilterQuery{
+		Topics:    warpFilterQuery.Topics,
+		Addresses: warpFilterQuery.Addresses,
+		FromBlock: height,
+	}
+	logs, err := ethClient.FilterLogs(context.Background(), initializationFilterQuery)
+	if err != nil {
+		s.logger.Error(
+			"Failed to get logs on initialization",
+			zap.Error(err),
+		)
+		return err
+	}
+
+	// Queue each of the logs to be processed
+	s.logger.Info(
+		"Processing logs on initialization",
+		zap.String("fromBlockHeight", height.String()),
+		zap.String("toBlockHeight", strconv.Itoa(int(latestBlock))),
+	)
+	for _, log := range logs {
+		messageInfo, err := NewWarpLogInfo(log)
+		if err != nil {
+			s.logger.Info(
```

Comment:
I think this log severity should be higher

Reply:
same for in `forwardLog`

Reply:
Good call. I also added errors logs to `NewWarpLogInfo`. Even though we're processing logs from an external source, this rises to the level of an error because those logs are pre-filtered in the `eth_getLogs` query such that we should always be able to create a `WarpLogInfo` from them.